lib/api/model/model.dart

```diff
@@ -531,10 +533,68 @@ String? tryParseEmojiCodeToUnicode(String emojiCode) {
   }
 }
 
+/// As in [MessageBase.recipient].
+///
+/// Different from [MessageDestination], this information comes from
+/// [getMessages] or [getEvents], identifying the conversation that contains a
+/// message.
+sealed class Recipient {}
+
+/// The recipient of a stream message.
+@JsonSerializable(fieldRename: FieldRename.snake, createToJson: false)
+class StreamRecipient extends Recipient {
+  StreamRecipient(this.streamId, this.topic);
+
+  int streamId;
+
+  @JsonKey(name: 'subject')
+  TopicName topic;
+
+  factory StreamRecipient.fromJson(Map<String, dynamic> json) =>
+    _$StreamRecipientFromJson(json);
+}
+
+/// The recipient of a DM message.
+class DmRecipient extends Recipient {
+  DmRecipient({required this.allRecipientIds})
+    : assert(isSortedWithoutDuplicates(allRecipientIds.toList()));
+
+  /// The user IDs of all users in the thread, sorted numerically.
+  ///
+  /// This lists the sender as well as all (other) recipients, and it
+  /// lists each user just once.  In particular the self-user is always
+  /// included.
+  ///
+  /// This is required to have an efficient `length`.
+  final List<int> allRecipientIds;
```

Comment:
```suggestion
  final List<int> allRecipientIds;
```

For a `List`, it's expected that `length` will be efficient.

(I guess the `List.length` doc doesn't say that. But lots and lots of code assumes it, because it's true for the built-in lists and more generally for any array data structure. Similarly lots of code assumes that `operator []` is efficient on a `List`.)